junit-jupiter-engine/src/test/kotlin/org/junit/jupiter/api/KotlinAssertionsTests.kt

```diff
@@ -211,6 +211,59 @@ class KotlinAssertionsTests {
         assertMessageStartsWith(error, assertionMessage)
     }
 
+    @Test
+    fun `assertNotNull with compiler smart cast`() {
+        val nullableString: String? = "string"
+
+        assertNotNull(nullableString)
+        assertFalse(nullableString.isEmpty()) // smart cast to a non nullable object
+    }
+
+    @Test
+    fun `assertNull with compiler smart cast`() {
+        val nullableString: String? = null
+
+        assertNull(nullableString)
+        // even safe call is not allowed, because compiler knows that string is always null
+        // nullableString?.isEmpty()
+    }
+
+    @Test
+    fun `assertInstanceOf with compiler smart cast`() {
+        val string: Any = "string"
+
+        assertInstanceOf<String>(string)
+        assertFalse(string.isEmpty()) // smart cast to a String object
+    }
+
+    @Test
+    fun `assertInstanceOf with compiler nullable smart cast`() {
+        val string: Any? = "string"
+
+        assertInstanceOf<String>(string)
+        assertFalse(string.isEmpty()) // smart cast to a non-null String object
+    }
+
+    @Test
+    fun `assertInstanceOf with a null value`() {
+        val error = assertThrows<AssertionFailedError> {
+            assertInstanceOf<String>(null)
+        }
```

Comment:
This feels meta, using a Kotlin assertion to test a Kotlin assertion, but it's probably ok, since they have different use cases for these two.

Reply:
We have a few tests for `assertThrows`. Also, a similar approach is used for `assertDoesNotThrow` tests